### PR TITLE
debug: coredump: Adds retained_memory as a backend

### PR DESCRIPTION
**Pull request by ldenefle**
This PR adds the possibility to use retained memory to store coredumps. 

It's very useful if the device does not embed a flash or if more complex processing is needed on the coredump.

It however has a couple drawbacks:
- Retained memory might be subject to power losses
- Locking mechanism on the retained memory device has to be turned off has mutexes don't work in ISR context.
- Can only store one coredump at a time, it's the responsibility of the app to read it and do something with it between different panics.

### CODE REVIEW

File: subsys/debug/coredump/Kconfig [hunk 1]
```diff
@@ -28,6 +28,14 @@ config DEBUG_COREDUMP_BACKEND_FLASH_PARTITION
 	  Core dump is saved to a flash partition with DTS alias
 	  "coredump-partition".
 
+config DEBUG_COREDUMP_BACKEND_RETAINED_MEM
+	bool "Use flash partition for coredump"
```
Comment (nordicjm) [inline subsys/debug/coredump/Kconfig:32]: Comment needs updating

File: subsys/debug/coredump/Kconfig [hunk 2]
```diff
@@ -28,6 +28,14 @@ config DEBUG_COREDUMP_BACKEND_FLASH_PARTITION
...
+	depends on RETAINED_MEM
+	help
+	  Core dump is saved to retained memory to be used
+	  next boot.
+
+
```
Comment (nordicjm) [inline subsys/debug/coredump/Kconfig:38]: Second newline can go

File: drivers/retained_mem/Kconfig [hunk 3]
```diff
@@ -15,6 +15,13 @@ config RETAINED_MEM_INIT_PRIORITY
 	help
 	  Retained memory devices initialization priority,
 
+config RETAINED_MEM_CONCURRENT_ACCESS
+	bool "If the retained mem will have concurrent access"
+    default y if MULTITHREADING
```
Comment (nordicjm) [inline drivers/retained_mem/Kconfig:20]: 1x tab for Kconfigs (help text is 1x tab followed 2x spaces)

File: subsys/debug/coredump/coredump_backend_retained_mem.c [hunk 4]
```diff
@@ -0,0 +1,190 @@
+#include <errno.h>
```
Comment (nordicjm) [inline subsys/debug/coredump/coredump_backend_retained_mem.c:1]: Copyright and license part needed

**Review by nordicjm — COMMENTED** [review]
Nice, didn't envisage someone would have a use for retained memory right away, that is a cool feature! Had a quick look at the files and added some comments.

**Comment by ldenefle** [comment]
The `retained_mem` subsystem comes in handy as on my current project, we have a lot of custom logging formatting so can't really use the `logging` backend and the `flash_partition` one is a bit overblown for my need. 
I really like the simplicity of retained memory over the other backends so it's perfect for my use case :+1:

File: subsys/debug/coredump/Kconfig [hunk 5]
```diff
@@ -28,6 +28,13 @@ config DEBUG_COREDUMP_BACKEND_FLASH_PARTITION
 	  Core dump is saved to a flash partition with DTS alias
 	  "coredump-partition".
 
+config DEBUG_COREDUMP_BACKEND_RETAINED_MEM
+	bool "Use retained mem subsys for coredump"
```
Comment (nordicjm) [inline subsys/debug/coredump/Kconfig:32]:
```suggestion
	bool "Use retained memory for coredump"
```
Retained memory is a driver, one important thing to note is that retained memory will take up a whole area, e.g. if you use a 10KB RAM area for it, it is the whole 10KB RAM area, you can't only use e.g. 2KB and then use 8KB for something else. The data retention subsystem builds on top of retained memory and does add this support, it also adds an optional checksum and optional magic header to the data area - it is currently in review at https://github.com/zephyrproject-rtos/zephyr/pull/55125

File: drivers/retained_mem/Kconfig [hunk 6]
```diff
@@ -15,6 +15,13 @@ config RETAINED_MEM_INIT_PRIORITY
 	help
 	  Retained memory devices initialization priority,
 
+config RETAINED_MEM_CONCURRENT_ACCESS
+	bool "If the retained mem will have concurrent access"
+	default y if MULTITHREADING
+	help
+	  Enables support for drivers that can retain their data whilst the
+	  device is powered (may be lost in low power states).
+
```
Comment (nordicjm) [inline drivers/retained_mem/Kconfig:24]: Comment needs updating

File: drivers/retained_mem/Kconfig [hunk 7]
```diff
@@ -15,6 +15,13 @@ config RETAINED_MEM_INIT_PRIORITY
 	help
 	  Retained memory devices initialization priority,
 
+config RETAINED_MEM_CONCURRENT_ACCESS
+	bool "If the retained mem will have concurrent access"
```
Comment (nordicjm) [inline drivers/retained_mem/Kconfig:19]:
```suggestion
	bool "Retained memory locking (mutex) support"
```

File: subsys/debug/coredump/coredump_backend_retained_mem.c [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+#include "coredump_internal.h"
+
+const static struct device *coredump_retained_mem_dev =
+	DEVICE_DT_GET(DT_ALIAS(coredump_retained_mem));
+static size_t written;
+static int error;
+
+struct retainem_mem_hdr_t {
+	/* 'C', 'D' */
+	char magic[2];
```
[the remark this answers is not in the document]
Comment (ldenefle) [reply subsys/debug/coredump/coredump_backend_retained_mem.c:20]: Sounds good, I could rebase my PR on top of the data retention branch

File: drivers/retained_mem/retained_mem_nrf_gpregret.c [hunk 9]
```diff
@@ -125,7 +125,7 @@ static const struct retained_mem_driver_api nrf_gpregret_api = {
 	DEVICE_DT_INST_DEFINE(inst,						\
 			      &nrf_gpregret_init,				\
 			      NULL,						\
-			      COND_CODE_1(CONFIG_MULTITHREADING,		\
+			      COND_CODE_1(CONFIG_RETAINED_MEM_CONCURRENT_ACCESS,		\
 					  (&nrf_gpregret_data_##inst), (NULL)	\
 			      ),						\
 			      &nrf_gpregret_config_##inst,			\
```
[the remark this answers is not in the document]
Comment (ldenefle) [reply drivers/retained_mem/retained_mem_nrf_gpregret.c:131]: My clang-format acts a bit crazy around those big macros, do you have any advice on how to deal with them ?

File: subsys/debug/coredump/Kconfig (hunk 5 above)
Comment (ldenefle) [reply subsys/debug/coredump/Kconfig:32]: any ideas on when #55125 should get merged ?

Comment (nordicjm) [reply subsys/debug/coredump/Kconfig:32]: Unfortunately not, it seems to be suffering from a lack of people that want to review it, feel free to review it though :)

File: drivers/retained_mem/retained_mem_nrf_gpregret.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (nordicjm) [reply drivers/retained_mem/retained_mem_nrf_gpregret.c:131]: Is this using the default zephyr .clang-format file? @gmarull

[the remark this answers is not in the document]
Comment (ldenefle) [reply drivers/retained_mem/retained_mem_nrf_gpregret.c:131]: Yep as far as I know

File: subsys/debug/coredump/Kconfig (hunk 5 above)
Comment (ldenefle) [reply subsys/debug/coredump/Kconfig:32]: Didn't realise I could review it, I'm not a maintainer

**Comment by ldenefle** [comment]
Just for clarity, this PR is currently being held up by #55125  once it's merged I'll rebase it and rewrite some bits to fit the retention subsys

**Comment by nordicjm** [comment]
> Just for clarity, this PR is currently being held up by #55125 once it's merged I'll rebase it and rewrite some bits to fit the retention subsys

Can you review https://github.com/zephyrproject-rtos/zephyr/pull/55125 and leave comments if you think something should change or approve if you think it looks okay? Will help in getting it merged.

**Comment by ldenefle** [comment]
Hello @nordicjm considering #55125 seem to be on hold, what would your recommendations be about this PR ?

**Comment by nordicjm** [comment]
It's on hold because I'm awaiting @d3zd3z to merge https://github.com/mcu-tools/mcuboot/pull/1629 in mcuboot

**Comment by ldenefle** [comment]
@nordicjm will do today

**Comment by ldenefle** [comment]
I'll rebase that onto the now upstream data_retention subsystem

**Comment by MaureenHelm** [comment]
@nashif please take a look

**Comment by Laczen** [comment]
@ldenefle, a recent addition to the flash simulator allows to assign a memory region to the flash simulator (PR #57380). You can use the available `coredump_backend_flash_partition` without the need to add an extra backend.

**Comment by ldenefle** [comment]
@Laczen that looks good, I think I can close this PR for the time being

**Pull request closed without merging**